clippy_lints/src/functions/ref_option.rs

```diff
@@ -0,0 +1,108 @@
+use crate::functions::hir::GenericArg;
+use crate::functions::REF_OPTION;
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::source::snippet;
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_hir::intravisit::FnKind;
+use rustc_hir::{FnDecl, GenericArgs, ImplItem, MutTy, QPath, TraitItem, Ty, TyKind};
+use rustc_lint::LateContext;
+use rustc_span::{sym, Span};
+
+fn check_ty(cx: &LateContext<'_>, param: &Ty<'_>, fixes: &mut Vec<(Span, String)>) {
+    // TODO: is this the right way to check if ty is an Option?
+    if let TyKind::Ref(lifetime, MutTy { ty, .. }) = param.kind
+        && let TyKind::Path(QPath::Resolved(_, path)) = ty.kind
+        && path.segments.len() == 1
+        && let seg = &path.segments[0]
+        && seg.ident.name == sym::Option
+        // check if option contains a regular type, not a reference
+        // TODO: Should this instead just check that opt_ty is a TyKind::Path?
+        && let Some(GenericArgs { args: [GenericArg::Type(opt_ty)], .. }) = seg.args
+        && !matches!(opt_ty.kind, TyKind::Ref(..))
+    {
+        // FIXME: Should this use the Option path from the original type?
+        // FIXME: Should reference be added in some other way to the snippet?
+        // FIXME: What should the lifetime be of the reference in Option<&T>?
```

Comment:
I think using `Option` instead of a full path is ok; it's in the prelude and people usually don't override it. If you want to be completely sure, you could use the `with_forced_trimmed_paths!` macro instead, but that's probably overkill here. Also we can take the lifetime of the original type (e.g. `&'x Option<T>` → `Option<&'x T>`), I don't think that should break the build.